### PR TITLE
feat(rum-core): end spa navigations after browser frame

### LINKED ISSUE #404: ending automatic transactions in a graceful way
Our transaction end detection logic for all transactions type except `page-load` is a bit eager now and might fail to capture spans that happen in the frames that happen after the nextTick. 

During the route-change events, there are multiple ways users would be invoking external API calls, can be sync, concurrent/parallel or even lazy load some components based on the API calls. Right now our framework integrations rely on the internal lifecycle events of the respective frameworks and tries to end the transaction as soon as the component is mounted to the DOM. 

With this approach, we might be able to only capture only the API calls that are requested during the component mounting and might fail to capture calls that are done after the component is mounted to the DOM. 

### Solution 
+ Instead of closing the transaction as soon as the children components are mounted or when the route change navigation is done, We could actively wait and listen for few frames (for now we can listen for 5 successive ticks/animation frames) and decide if we want to close the transaction. 
+ If there are no calls within those 5 ticks, we close the transaction immediately and send it to the APM server. 
+ If there are any spans, assume a call happens on the 3rd frame, then we schedule another 5 frames and end the transaction accordingly. 
+ In order to reduce the noise in the transaction, our transaction would always get time shifted to the last active transaction. Ex - if there are 2 spans that happens in the 2nd tick after the transaction start, we end the transaction with the end time of the longest span and send it to the server. 
+ With our filtering logic, transaction without any spans will not be reported to the server.

### PR DESCRIPTION
**Pull request by vigneshshanmugam**
+ Alternative solution to #729 
+ fixes #404 #636 
+ The solution is based on scheduling the transaction to be ended after the browser paints the new frame ( includes script, layout, style and paint). This ensures that if there is any network activity or if the underlying framework yields to the browser to do the work (style, paint). We include those timings as well which would help us measure most of the layout changes. 

There is a detailed [blog post ](https://nolanlawson.com/2018/09/25/accurately-measuring-layout-on-the-web/)explaining different ways to schedule a call after next frame. We are using the `rAF` + `setTimeout` technique as we want the call to get executed even if the tab is opened in invisible mode (cmd + click). 

### Pros
+ Simpler and Easier to reason about, don't need to keep the state on the frame running and scheduling problems
+ Gets called even if the current tab is invisible but with delays (due to the use of timers). `RAF` never gets called for invisible tabs. 

### Cons
+ We wont be able to measure if the user does `setTimeout(task, 1000)` since it could be scheduled by the browser in next to next frames and we only measure the next frame after the navigation. This should be okay as it is not a usual thing. 
+ we might increase the transaction duration. by `16.67ms` worst case due to the use of rAF, but we would be able to capture something the frameworks/user is doing during layout so might be worth it. 

Personal Opinion, I like this solution better than the other one due to its simplicity and solving for 99% of the use cases. Was able to work for most of the cases when the call is scheduled in
`setImmediate`, `setTimeout(call, 0)`, `requestAnimationFrame`, `requestIdleCallback`, `Promise.resolve`, `MessageChannel` during manual testing.

### CODE REVIEW

**Comment by jahtalab** [comment]
- test for `afterFrame` to investigate behaviour on different browsers
- We also need test for react.lazy

**Comment by vigneshshanmugam** [comment]
jenkins run the tests please

**Comment by vigneshshanmugam** [comment]
Chrome 76 seems to have an issue with scheduling tasks. test passes on all browsers except for this one.

File: packages/rum-react/test/e2e/with-router/switch.e2e-spec.js [hunk 1]
```diff
@@ -59,4 +58,32 @@ describe('Using Switch component of react router', function() {
...
+
+    const pageLoadTransaction = transactions[0]
+    expect(pageLoadTransaction.type).toBe('page-load')
+    expect(pageLoadTransaction.name).toBe('/notfound')
+
+    const routeTransaction = transactions[1]
+    expect(routeTransaction.name).toBe('/lazy')
+    expect(routeTransaction.type).toBe('route-change')
+
+    expect(routeTransaction.spans.length).toBe(1)
```
Comment (jahtalab) [inline packages/rum-react/test/e2e/with-router/switch.e2e-spec.js:86]: Shouldn't the request that fetches the separate JS bundle be captured from the resource timing?

**Comment by vigneshshanmugam** [comment]
@jahtalab 
> Would you please comment about the Chrome 76 scheduling issue here?

Its not chrome 76 scheduling issue, I added more details with comments in the test itself. It's hard to know test this scenario for different browsers, thats why the test checks for 2 or 3 more, Also its not a common scenario - No one of the frameworks that I have tested(React, Vue, Angular, Preact) use this double timeouts. 

There is no underlying issue with our approach, instead if the user really uses this way of rendering something, Neither the frameworks not our library would be able to track it. Unless we monkey patch.

File: packages/rum-core/test/common/after-frame.spec.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      attributes: true
+    })
+    // Trigger mutation observer callback
+    document.body.setAttribute('test', 'test')
+  })
+
+  it('should be called after tasks', done => {
+    let count = 0
+    afterFrame(() => {
+      expect(count).toBeGreaterThanOrEqual(2)
```
[the remark this answers is not in the document]
Comment (vigneshshanmugam) [reply packages/rum-core/test/common/after-frame.spec.js:47]: There is a comment below that explains it. I can move this the `double timeout` test as a edge case test and add comments.

File: packages/rum-react/test/e2e/with-router/switch.e2e-spec.js (hunk 1 above)
Comment (vigneshshanmugam) [reply packages/rum-react/test/e2e/with-router/switch.e2e-spec.js:86]: It wont, since use the pre bundled functional component, I will create a new lazy component and add it.

**Comment by vigneshshanmugam** [comment]
Wrote a detailed comment on the double setTimeout tests https://github.com/elastic/apm-agent-rum-js/pull/730/files#diff-c6439d182e960bb8b3e580ba6c8e250aR61-R88 and also moved them to edge case test.

**Comment by vigneshshanmugam** [comment]
Lazy component mount cannot be solved with this PR as our Route component wont be even kicked in(Transaction will not start until React fetches all underlying dependencies) when the dependencies are being fixed.  Also this PR is not for addressing that issue, instead measuring everything that happens before the browser renders the next frame in a better way. 

However I added tests when lazy components are fetched inside a component that is child of our  `ApmRoute` component. 

Lazy components can be still measured effectively by the technique already mentioned here - https://www.elastic.co/guide/en/apm/agent/rum-js/current/react-integration.html#_instrumenting_lazy_loaded_routes
